extension/src/extension.ts

```diff
@@ -4,17 +4,19 @@ import { HistoryDebuggerPanel } from "./panel"
 import { Server } from "./server"
 
 export async function activate(context: vscode.ExtensionContext): Promise<void> {
-  // TODO: destroy server if panel is destroyed?
+  const debugAdapterTrackerFactory: vscode.Disposable = registerDebugAdapterTrackerFactory()
   let server: Server | undefined = undefined
-  context.subscriptions.push(
-    vscode.commands.registerCommand("temporal.debugger-extension.open-panel", async () => {
-      if (!server) {
-        server = await Server.create()
-        console.log(`Server listening on ${server.url}`)
+  const openCommand: vscode.Disposable = vscode.commands.registerCommand(
+    "temporal.debugger-extension.open-panel",
+    async () => {
+      if (server) {
+        // close existing server if command is executed while panel is still open
+        server.terminate()
```

Comment:
I see what you did here but I'm not convinced it's the right approach, it seems racy and spreads the server lifecycle management across the codebase.
I'd create a new server instance per panel instead of storing this reusable reference.

Reply:
This is still creating a new server instance per panel. The edge case is when someone already has the panel open and then selects the open command again. The `onDidDispose` isn't triggered so the previous server isn't closed, but we open a new server and a new instance of the panel.

Maybe a better solution is to dispose of the panel or just check if there is an existing panel first. I'll think on it a bit more 🤔